### PR TITLE
Construct question id based on nickname and question number

### PR DESCRIPTION
**Pull request by TylerZeroMaster**
fixes openstax/ce#2226

Question number is only used when applicable (i.e. in question sets)

If the exercise is a question set and where `n` and `i` are the respective indices, the id takes the form: 
* `{nickname}_question{n}` for questions
* ~`{nickname}_question{n}_answer{i}` for answers~

If the exercise is not a question set and where `n` and `i` are the respective indices, the id takes the form:
* `{nickname}` for questions
* ~`{nickname}_answer{i}` for answers~

Additionally, instead of setting `data-id`, the template now sets `id`. For legacy exercises, the ids should match what had been generated by cookbook. Cookbook only tried to generate an id if the id was not already set; therefore, setting the id here should alleviate the issue in cookbook.

- The format for question ids is `auto_{parent_page_uuid}_{question.id}`
- ~The format for answer ids is `auto_{parent_page_uuid}_{answer.id}`~ answer ids removed (they were not being used)

### CODE REVIEW

**Review by philschatz — CHANGES_REQUESTED** [review]
Any reason this concatenation isn't done in the cookbook? I wonder because there isn't a schema/validation that requires the `data-id=""` attribute to be unique within a document and I'm guessing the uniqueness is something cookbook relies on.

If cookbook really needs a string to be unique across the whole book (e.g. to move the answers to the end of the book and link to them) then what if we this concatenated identifier was on an `id=""` attribute instead?

File: nebuchadnezzar/nebu/h5p_injection.py [hunk 1]
```diff
@@ -218,16 +229,22 @@ def _add_question(
             assert (
                 sub_library != "H5P.QuestionSet"
             ), "Question sets cannot contain question sets"
-            _add_question(i + 1, sub_library, sub_entry, questions)
+            _add_question(
+                [*id_parts, f"q-{i + 1}"], sub_library, sub_entry, questions
```
Comment (philschatz) [inline nebuchadnezzar/nebu/h5p_injection.py:233]: Spelling out "question" might match with the use of `exercise_` earlier in the identifier and make it easier to find this line where the string is generated.

Also, why do underscore in `exercise_` but hyphen here?

**Review by philschatz — COMMENTED** [review]
Any reason this concatenation isn't done in the cookbook? I wonder because there isn't a schema/validation that requires the `data-id=""` attribute to be unique within a document and I'm guessing the uniqueness is something cookbook relies on.

If cookbook really needs a string to be unique across the whole book (e.g. to move the answers to the end of the book and link to them) then what if we this concatenated identifier was on an `id=""` attribute instead?

**Comment by TylerZeroMaster** [comment]
> Any reason this concatenation isn't done in the cookbook? I wonder because there isn't a schema/validation that requires the `data-id=""` attribute to be unique within a document and I'm guessing the uniqueness is something cookbook relies on.
> 
> If cookbook really needs a string to be unique across the whole book (e.g. to move the answers to the end of the book and link to them) then what if we this concatenated identifier was on an `id=""` attribute instead?

Yes, in the issue it is explained that modifying this in cookbook is not a good idea because (1) the error is created when **generating** the id and (2) it would change **all** the ids for existing exercises too. The error from cookbook is related to trying to find the element's parent page and it occurs any time you try to find the parent page (not just when generating the id). As stated in the issue, this problem exists for current/legacy exercises as well if you insert the same exercise on the same page multiple times.

Adding an id directly during assembly could work.

File: nebuchadnezzar/nebu/h5p_injection.py (hunk 1 above)
Comment (TylerZeroMaster) [reply nebuchadnezzar/nebu/h5p_injection.py]: I was thinking that `_` was used to separate parts of the id and `-` was used to join segments within a part (like in a uuid, for example). I removed the hyphens in favor of adding the number directly to the end (i.e. question1 or answer1).

**Review by philschatz — APPROVED** [review]
:+1:

**Pull request merged**